src/data/dfinsupp/basic.lean

```diff
@@ -1040,6 +1044,121 @@ assume f g, decidable_of_iff (f.support = g.support ∧ (∀i∈f.support, f i =
         by rw [hf, hg],
     by intro h; subst h; simp⟩
 
+section equiv
+open finset
+variables [Π i, has_zero (β i)]
+
+variables {κ : Type*} --[decidable_eq κ] --[Π i (x : β i), decidable (x ≠ 0)]
+/--Reindexing (and possibly removing) terms of a dfinsupp.-/
+noncomputable def comap_domain (h : κ → ι) (hh : function.injective h) :
+(Π₀ i, β i) → Π₀ k, β (h k) :=
+begin
+  refine quotient.lift (λ f, ⟦_⟧) (λ f f' h, _),
+  exact { to_fun := λ x, f.to_fun (h x),
+    pre_support := (f.pre_support.to_finset.preimage h (hh.inj_on _)).val,
+    zero := λ x, (f.zero (h x)).imp_left $ λ hx, mem_preimage.mpr $ multiset.mem_to_finset.mpr hx },
+  exact quot.sound (λ x, h _)
+end
+@[simp] lemma comap_domain_apply (h : κ → ι) (hh : function.injective h) (f : Π₀ i, β i) (k : κ) :
+comap_domain h hh f k = f (h k) :=
+by { rcases f, refl }
+
+omit dec
+/--A computable version of comap_domain when an explicit left inverse is provided.-/
+def comap_domain' (h : κ → ι) {h' : ι → κ} (hh' : function.left_inverse h' h) :
+  (Π₀ i, β i) → (Π₀ k, β (h k)) :=
+begin
+  refine quotient.lift (λ f, ⟦_⟧) (λ f f' h, _),
+  exact { to_fun := λ x, f.to_fun (h x),
+    pre_support := f.pre_support.map h',
+    zero := λ x, (f.zero (h x)).imp_left $ λ hx, multiset.mem_map.mpr ⟨_, hx, hh' _⟩ },
+  exact quot.sound (λ x, h _),
+end
+@[simp] lemma comap_domain'_apply (h : κ → ι) {h' : ι → κ} (hh' : function.left_inverse h' h)
+(f : Π₀ i, β i) (k : κ) : comap_domain' h hh' f k = f (h k) :=
+by { rcases f, refl }
+
+/--Reindexing terms of a dfinsupp.-/
+def equiv_congr_left (h : ι ≃ κ) : (Π₀ i, β i) ≃ (Π₀ k, β (h.symm k)) :=
+⟨comap_domain' h.symm h.right_inv,
+λ f, map_range (λ i, equiv.cast $ congr_arg β $ h.symm_apply_apply i)
+  (λ i, (equiv.cast_eq_iff_heq _).mpr $
+    by { convert heq.rfl, repeat { exact (h.symm_apply_apply i).symm } })
+  (@comap_domain' _ _ _ _ h _ h.left_inv f),
+λ f, by { ext i, rw [map_range_apply, comap_domain'_apply, comap_domain'_apply,
+  equiv.cast_eq_iff_heq, h.symm_apply_apply] },
+λ f, by { ext k, rw [comap_domain'_apply, map_range_apply, comap_domain'_apply,
+  equiv.cast_eq_iff_heq, h.apply_symm_apply] }⟩
+
+@[simp] lemma equiv_congr_left_apply (h : ι ≃ κ) (f : Π₀ i, β i) (k : κ) :
+equiv_congr_left h f k = f (h.symm k) := comap_domain'_apply h.symm h.right_inv f k
+
+section curry
+variables {α : ι → Type*} {δ : (Σ i, α i) → Type v} [Π i, has_zero (δ i)]
+
+/--The natural map between `Π₀ (i : Σ i, α i), δ i` and `Π₀ i (j : α i), δ ⟨i, j⟩`.-/
+noncomputable def curry (f : Π₀ i, δ i) : Π₀ i j, δ ⟨i, j⟩ := by { classical, exact
```

Comment:
I think this should be called `sigma_curry`, we normally reserve `curry` for the `prod` version. Note we have `sigma.curry` as the function version of this.

Reply:
Looking at `sigma.curry` suggests this would be more consistent with `δ` defined differently, as
```suggestion
noncomputable def curry (f : Π₀ i, δ i.fst I.snd) : Π₀ i j, δ i j := by { classical, exact
```
Perhaps you found this harder to work with? If so, can you add something to the docstring explaining the choice you made?